File: Sources/Features/Dog/BreedDetailViewController.m
#import "BreedDetailViewController.h"
#import "BreedDetailViewModel.h"
#import "ImageLoader.h"

#pragma mark - Image Cell

@interface DogImageCell : UICollectionViewCell
@property (nonatomic, strong) UIImageView *imageView;
@property (nonatomic, copy) NSString *representedURL;
@end

@implementation DogImageCell
- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        _imageView = [[UIImageView alloc] initWithFrame:self.contentView.bounds];
        _imageView.contentMode = UIViewContentModeScaleAspectFill;
        _imageView.clipsToBounds = YES;
        _imageView.backgroundColor = [UIColor secondarySystemBackgroundColor];
        _imageView.layer.cornerRadius = 8;
        _imageView.autoresizingMask =
            UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        [self.contentView addSubview:_imageView];
    }
    return self;
}
- (void)prepareForReuse {
    [super prepareForReuse];
    self.imageView.image = nil;
    self.representedURL = nil;
}
@end

#pragma mark - Header

@interface DogDetailHeaderView : UICollectionReusableView
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UILabel *subtitleLabel;
@end

@implementation DogDetailHeaderView
- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        _titleLabel = [[UILabel alloc] init];
        _titleLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleTitle1];
        _titleLabel.translatesAutoresizingMaskIntoConstraints = NO;

        _subtitleLabel = [[UILabel alloc] init];
        _subtitleLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];
        _subtitleLabel.textColor = [UIColor secondaryLabelColor];
        _subtitleLabel.numberOfLines = 0;
        _subtitleLabel.translatesAutoresizingMaskIntoConstraints = NO;

        [self addSubview:_titleLabel];
        [self addSubview:_subtitleLabel];
        [NSLayoutConstraint activateConstraints:@[
            [_titleLabel.topAnchor constraintEqualToAnchor:self.topAnchor constant:12],
            [_titleLabel.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
            [_titleLabel.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
            [_subtitleLabel.topAnchor constraintEqualToAnchor:_titleLabel.bottomAnchor constant:6],
            [_subtitleLabel.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
            [_subtitleLabel.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
            [_subtitleLabel.bottomAnchor constraintLessThanOrEqualToAnchor:self.bottomAnchor constant:-8],
        ]];
    }
    return self;
}
@end

#pragma mark - View Controller

static NSString *const kImageCellID = @"DogImageCell";
static NSString *const kHeaderID = @"DogDetailHeaderView";

@interface BreedDetailViewController () <UICollectionViewDataSource, UICollectionViewDelegateFlowLayout>
@property (nonatomic, strong) BreedDetailViewModel *viewModel;
@property (nonatomic, strong) UICollectionView *collectionView;
@property (nonatomic, strong) UIActivityIndicatorView *spinner;
@property (nonatomic, strong) UIBarButtonItem *favoriteButton;
@end

@implementation BreedDetailViewController

- (instancetype)initWithBreed:(Breed *)breed {
    self = [super initWithNibName:nil bundle:nil];
    if (self) {
        _viewModel = [[BreedDetailViewModel alloc] initWithBreed:breed];
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = self.viewModel.title;
    self.view.backgroundColor = [UIColor systemBackgroundColor];
    [self setupCollectionView];
    [self setupSpinner];
    [self setupFavoriteButton];
    [self bindViewModel];
    [self.viewModel loadImages];
}

- (void)setupCollectionView {
    UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc] init];
    layout.minimumInteritemSpacing = 8;
    layout.minimumLineSpacing = 8;
    layout.sectionInset = UIEdgeInsetsMake(8, 16, 16, 16);

    self.collectionView = [[UICollectionView alloc] initWithFrame:CGRectZero
                                             collectionViewLayout:layout];
    self.collectionView.translatesAutoresizingMaskIntoConstraints = NO;
    self.collectionView.backgroundColor = [UIColor systemBackgroundColor];
    self.collectionView.dataSource = self;
    self.collectionView.delegate = self;
    [self.collectionView registerClass:[DogImageCell class]
            forCellWithReuseIdentifier:kImageCellID];
    [self.collectionView registerClass:[DogDetailHeaderView class]
            forSupplementaryViewOfKind:UICollectionElementKindSectionHeader
                   withReuseIdentifier:kHeaderID];

    [self.view addSubview:self.collectionView];
    [NSLayoutConstraint activateConstraints:@[
        [self.collectionView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
        [self.collectionView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor],
        [self.collectionView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [self.collectionView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
    ]];
}

- (void)setupSpinner {
    self.spinner = [[UIActivityIndicatorView alloc]
        initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleLarge];
    self.spinner.hidesWhenStopped = YES;
    self.spinner.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.spinner];
    [NSLayoutConstraint activateConstraints:@[
        [self.spinner.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [self.spinner.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor],
    ]];
}

- (void)setupFavoriteButton {
    self.favoriteButton =
        [[UIBarButtonItem alloc] initWithImage:nil
                                         style:UIBarButtonItemStylePlain
                                        target:self
                                        action:@selector(favoriteTapped)];
    self.navigationItem.rightBarButtonItem = self.favoriteButton;
    [self updateFavoriteButton];
}

- (void)bindViewModel {
    __weak typeof(self) weakSelf = self;
    self.viewModel.onStateChange = ^{
        __strong typeof(weakSelf) self = weakSelf;
        if (!self) { return; }
        if (self.viewModel.isLoading) {
            [self.spinner startAnimating];
        } else {
            [self.spinner stopAnimating];
        }
        [self updateFavoriteButton];
        [self.collectionView reloadData];
    };
}

- (void)updateFavoriteButton {
    BOOL isFav = [self.viewModel isFavorite];
    NSString *symbol = isFav ? @"heart.fill" : @"heart";
    self.favoriteButton.image = [UIImage systemImageNamed:symbol];
    self.favoriteButton.tintColor =
        isFav ? [UIColor systemRedColor] : nil;
}

- (void)favoriteTapped {
    [self.viewModel toggleFavorite];
    // 状態は通知経由で onStateChange に反映される
}

#pragma mark - UICollectionViewDataSource

- (NSInteger)collectionView:(UICollectionView *)collectionView
     numberOfItemsInSection:(NSInteger)section {
    return self.viewModel.imageURLs.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView
                  cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    DogImageCell *cell =
        [collectionView dequeueReusableCellWithReuseIdentifier:kImageCellID
                                                  forIndexPath:indexPath];
    NSString *urlString = self.viewModel.imageURLs[indexPath.item];
    cell.representedURL = urlString;
    [[ImageLoader sharedLoader] loadImageURL:urlString
                                  completion:^(UIImage *image) {
        // セルが別の URL に再利用されていないか確認
        if ([cell.representedURL isEqualToString:urlString]) {
            cell.imageView.image = image;
        }
    }];
    return cell;
}

- (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView
           viewForSupplementaryElementOfKind:(NSString *)kind
                                 atIndexPath:(NSIndexPath *)indexPath {
    DogDetailHeaderView *header =
        [collectionView dequeueReusableSupplementaryViewOfKind:kind
                                           withReuseIdentifier:kHeaderID
                                                  forIndexPath:indexPath];
    header.titleLabel.text = self.viewModel.breed.displayName;
    NSString *subs = self.viewModel.breed.subBreedsDescription;
    header.subtitleLabel.text =
        subs ? [NSString stringWithFormat:@"サブ犬種: %@", subs] : @"サブ犬種なし";
    return header;
}

#pragma mark - UICollectionViewDelegateFlowLayout

- (CGSize)collectionView:(UICollectionView *)collectionView
                  layout:(UICollectionViewLayout *)collectionViewLayout
  sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    CGFloat available = collectionView.bounds.size.width - 16 * 2 - 8;
    CGFloat side = floor(available / 2.0);
    return CGSizeMake(side, side);
}

- (CGSize)collectionView:(UICollectionView *)collectionView
                  layout:(UICollectionViewLayout *)collectionViewLayout
referenceSizeForHeaderInSection:(NSInteger)section {
    return CGSizeMake(collectionView.bounds.size.width, 90);
}

@end

File: Sources/Features/Dog/BreedDetail/Presentation/BreedDetailViewController.swift
import UIKit

/// 犬種詳細画面。Dog タブ・お気に入りタブの双方から使用する（ObjC からも参照される）。
@objc(BreedDetailViewController)
final class BreedDetailViewController: UIViewController {

    private static let imageCellID = "DogImageCell"
    private static let headerID = "DogDetailHeaderView"

    private let viewModel: BreedDetailViewModel
    private var collectionView: UICollectionView!
    private var spinner: UIActivityIndicatorView!
    private var favoriteButton: UIBarButtonItem!

    // MARK: - Init

    @objc init(breed: Breed) {
        let item = BreedItem(
            name: breed.name,
            displayName: breed.displayName,
            subBreedsDescription: breed.subBreedsDescription
        )
        self.viewModel = BreedDetailViewModel(
            breed: item,
            fetchImagesUseCase: DefaultFetchBreedImagesUseCase(
                repository: DogAPIBreedImagesRepository()
            ),
            favoritesUseCase: DefaultFavoritesUseCase(
                repository: DefaultsFavoritesRepository()
            )
        )
        super.init(nibName: nil, bundle: nil)
    }

    /// Designated init から派生（テスト・DI 用途）。
    init(viewModel: BreedDetailViewModel) {
        self.viewModel = viewModel
        super.init(nibName: nil, bundle: nil)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) is not supported")
    }

    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()
        title = viewModel.title
        view.backgroundColor = .systemBackground
        setupCollectionView()
        setupSpinner()
        setupFavoriteButton()
        bindViewModel()
        viewModel.loadImages()
    }

    // MARK: - Setup

    private func setupCollectionView() {
        let layout = UICollectionViewFlowLayout()
        layout.minimumInteritemSpacing = 8
        layout.minimumLineSpacing = 8
        layout.sectionInset = UIEdgeInsets(top: 8, left: 16, bottom: 16, right: 16)

        collectionView = UICollectionView(frame: .zero, collectionViewLayout: layout)
        collectionView.translatesAutoresizingMaskIntoConstraints = false
        collectionView.backgroundColor = .systemBackground
        collectionView.dataSource = self
        collectionView.delegate = self
        collectionView.register(DogImageCell.self,
                                forCellWithReuseIdentifier: Self.imageCellID)
        collectionView.register(
            DogDetailHeaderView.self,
            forSupplementaryViewOfKind: UICollectionView.elementKindSectionHeader,
            withReuseIdentifier: Self.headerID
        )

        view.addSubview(collectionView)
        NSLayoutConstraint.activate([
            collectionView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            collectionView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
            collectionView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            collectionView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
        ])
    }

    private func setupSpinner() {
        spinner = UIActivityIndicatorView(style: .large)
        spinner.hidesWhenStopped = true
        spinner.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(spinner)
        NSLayoutConstraint.activate([
            spinner.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            spinner.centerYAnchor.constraint(equalTo: view.centerYAnchor),
        ])
    }

    private func setupFavoriteButton() {
        favoriteButton = UIBarButtonItem(
            image: nil,
            style: .plain,
            target: self,
            action: #selector(favoriteTapped)
        )
        navigationItem.rightBarButtonItem = favoriteButton
        updateFavoriteButton()
    }

    private func bindViewModel() {
        viewModel.onStateChange = { [weak self] in
            guard let self else { return }
            if self.viewModel.isLoading {
                self.spinner.startAnimating()
            } else {
                self.spinner.stopAnimating()
            }
            self.updateFavoriteButton()
            self.collectionView.reloadData()
        }
    }

    private func updateFavoriteButton() {
        let isFav = viewModel.isFavorite()
        let symbol = isFav ? "heart.fill" : "heart"
        favoriteButton.image = UIImage(systemName: symbol)
        favoriteButton.tintColor = isFav ? .systemRed : nil
    }

    @objc private func favoriteTapped() {
        viewModel.toggleFavorite()
    }
}

// MARK: - UICollectionViewDataSource

extension BreedDetailViewController: UICollectionViewDataSource {
    func collectionView(_ collectionView: UICollectionView,
                        numberOfItemsInSection section: Int) -> Int {
        viewModel.imageURLs.count
    }

    func collectionView(_ collectionView: UICollectionView,
                        cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(
            withReuseIdentifier: Self.imageCellID,
            for: indexPath
        ) as! DogImageCell
        let url = viewModel.imageURLs[indexPath.item]
        let urlString = url.absoluteString
        cell.representedURL = urlString
        ImageLoader.shared().loadImageURL(urlString) { [weak cell] image in
            // セルが別の URL に再利用されていないかを確認
            guard let cell, cell.representedURL == urlString else { return }
            cell.imageView.image = image
        }
        return cell
    }

    func collectionView(_ collectionView: UICollectionView,
                        viewForSupplementaryElementOfKind kind: String,
                        at indexPath: IndexPath) -> UICollectionReusableView {
        let header = collectionView.dequeueReusableSupplementaryView(
            ofKind: kind,
            withReuseIdentifier: Self.headerID,
            for: indexPath
        ) as! DogDetailHeaderView
        header.titleLabel.text = viewModel.breed.displayName
        if let subs = viewModel.breed.subBreedsDescription {
            header.subtitleLabel.text = "サブ犬種: \(subs)"
        } else {
            header.subtitleLabel.text = "サブ犬種なし"
        }
        return header
    }
}

// MARK: - UICollectionViewDelegateFlowLayout

extension BreedDetailViewController: UICollectionViewDelegateFlowLayout {
    func collectionView(_ collectionView: UICollectionView,
                        layout collectionViewLayout: UICollectionViewLayout,
                        sizeForItemAt indexPath: IndexPath) -> CGSize {
        let available = collectionView.bounds.width - 16 * 2 - 8
        let side = floor(available / 2.0)
        return CGSize(width: side, height: side)
    }

    func collectionView(_ collectionView: UICollectionView,
                        layout collectionViewLayout: UICollectionViewLayout,
                        referenceSizeForHeaderInSection section: Int) -> CGSize {
        CGSize(width: collectionView.bounds.width, height: 90)
    }
}

// MARK: - Image Cell

private final class DogImageCell: UICollectionViewCell {
    let imageView = UIImageView()
    var representedURL: String?

    override init(frame: CGRect) {
        super.init(frame: frame)
        imageView.frame = contentView.bounds
        imageView.contentMode = .scaleAspectFill
        imageView.clipsToBounds = true
        imageView.backgroundColor = .secondarySystemBackground
        imageView.layer.cornerRadius = 8
        imageView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        contentView.addSubview(imageView)
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) is not supported")
    }

    override func prepareForReuse() {
        super.prepareForReuse()
        imageView.image = nil
        representedURL = nil
    }
}

// MARK: - Header

private final class DogDetailHeaderView: UICollectionReusableView {
    let titleLabel = UILabel()
    let subtitleLabel = UILabel()

    override init(frame: CGRect) {
        super.init(frame: frame)
        titleLabel.font = .preferredFont(forTextStyle: .title1)
        titleLabel.translatesAutoresizingMaskIntoConstraints = false

        subtitleLabel.font = .preferredFont(forTextStyle: .body)
        subtitleLabel.textColor = .secondaryLabel
        subtitleLabel.numberOfLines = 0
        subtitleLabel.translatesAutoresizingMaskIntoConstraints = false

        addSubview(titleLabel)
        addSubview(subtitleLabel)
        NSLayoutConstraint.activate([
            titleLabel.topAnchor.constraint(equalTo: topAnchor, constant: 12),
            titleLabel.leadingAnchor.constraint(equalTo: leadingAnchor),
            titleLabel.trailingAnchor.constraint(equalTo: trailingAnchor),
            subtitleLabel.topAnchor.constraint(equalTo: titleLabel.bottomAnchor, constant: 6),
            subtitleLabel.leadingAnchor.constraint(equalTo: leadingAnchor),
            subtitleLabel.trailingAnchor.constraint(equalTo: trailingAnchor),
            subtitleLabel.bottomAnchor.constraint(lessThanOrEqualTo: bottomAnchor, constant: -8),
        ])
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) is not supported")
    }
}
